cedar-policy-core/src/ast/policy_set.rs

```diff
@@ -181,12 +298,58 @@ impl PolicySet {
                 id: oentry.key().clone(),
             }),
             Entry::Vacant(ventry) => {
+                self.template_to_links_map
+                    .insert(t.id().clone(), HashSet::new());
                 ventry.insert(Arc::new(t));
                 Ok(())
             }
         }
     }
 
+    /// Remove a template from the policy set.
+    /// If any policy is linked to the template, this will error
+    pub fn remove_template(
+        &mut self,
+        policy_id: &PolicyID,
+    ) -> Result<Template, PolicySetTemplateRemovalError> {
+        match self.template_to_links_map.get(policy_id) {
+            Some(map) => {
+                if !map.is_empty() {
+                    return Err(PolicySetTemplateRemovalError::RemoveTemplateWithLinksError(
+                        policy_id.clone(),
+                    ));
+                }
+            }
+            None => {
+                return Err(PolicySetTemplateRemovalError::RemovePolicyNoTemplateError(
+                    policy_id.clone(),
+                ))
+            }
+        };
+
+        // PANIC SAFETY: every linked policy should have a template
+        #[allow(clippy::panic)]
+        match self.templates.remove(policy_id) {
+            Some(t) => {
+                self.template_to_links_map.remove(policy_id);
+                Ok((*t).clone())
+            }
+            None => panic!("Found in template_to_links_map but not in templates"),
+        }
+    }
+
+    /// Get the list of policies linked to `template_id`.
+    /// Returns all p in `links` s.t. `p.template().id() == template_id`
+    pub fn get_linked_policies(
+        &mut self,
+        template_id: &PolicyID,
+    ) -> Result<Vec<&PolicyID>, PolicySetGetLinksError> {
```

Comment:
instead of `Vec<&PolicyID>`, we could slightly more flexibly/performantly return `impl Iterator<Item = &PolicyID>`; this means we wouldn't have to `collect()` (and thus allocate) in the `Ok` case